File: Day 065/Problem 139.Word Break.cpp
class Solution {
public:
    unordered_map<int, bool> dp;
    bool solve(int index, const string &s, const unordered_map<string, bool> &dict) {
        if (index == s.size()) {
            return true;
        }

        if (dp.find(index) != dp.end()) {
            return dp[index];
        }

        for (int i = index; i < s.size(); i++) {
            string str = s.substr(index, i - index + 1);
            if (dict.find(str) != dict.end()) {
                if (solve(i + 1, s, dict)) {
                    return dp[index] = true;
                }
            }
        }

        return dp[index] = false;
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_map<string, bool> dict;
        for (string word : wordDict) {
            dict[word] = true;
        }

        return solve(0, s, dict);
    }
};

File: Day 093/Problem 139. Word Break.java
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {

    Boolean[] dp;
    public boolean solve(int index, String s, Set<String> wordSet) {
        if (index == s.length()) {
            return dp[index] = true;
        }
        
        if (dp[index] != null) {
            return dp[index];
        }

        for (int i = index + 1; i <= s.length(); i++) {
            String substr = s.substring(index, i);
            if (wordSet.contains(substr) && solve(i, s, wordSet)) {
                return dp[index] = true;
            }
        }

        return dp[index] = false;
    }

    public boolean wordBreak(String s, List<String> wordDict) {
        dp = new Boolean[s.length() + 1];
        Set<String> wordSet = new HashSet<>(wordDict);
        return solve(0, s, wordSet);
    }
}